### PR TITLE
Fix service client thread leak

### PR DESCRIPTION
**Pull request by workingenius**
For cause of this problem, see [the first comment](https://github.com/turingvideo/rosjava_core/pull/5/files#r834896187)

### CODE REVIEW

File: rosjava/src/main/java/org/ros/internal/node/service/DefaultServiceClient.java [hunk 1]
```diff
@@ -119,11 +120,8 @@ public DefaultServiceClient(final GraphName nodeName,
     connectionHeader.addField(ConnectionHeaderFields.PERSISTENT, "1");
     connectionHeader.merge(serviceDeclaration.toConnectionHeader());
     tcpClientManager = new TcpClientManager(executorService);
-    final ServiceClientHandshakeHandler<T, S> serviceClientHandshakeHandler =
-        new ServiceClientHandshakeHandler<T, S>(connectionHeader, responseListeners, deserializer,
+    serviceClientHandshakeHandler = new ServiceClientHandshakeHandler<T, S>(connectionHeader, responseListeners, deserializer,
             executorService);
-    handshakeLatch = new HandshakeLatch();
-    serviceClientHandshakeHandler.addListener(handshakeLatch);
```
Comment (workingenius) [inline rosjava/src/main/java/org/ros/internal/node/service/DefaultServiceClient.java:126]: Each time this line is touched, a thread is leaked.
A new thread is created to wait for the coming handshake response, but they are never released no matter the handshake is successful or not. As a result, there's always be a new zombie thread after each service client creation.

In ordinary cases, there are only a small number of service clients, so this problem is not serious.
As we did service clients cache and recreation to simulate non-persist service calling, it explodes when there are many failure and many clients are created.

File: rosjava/src/main/java/org/ros/internal/node/service/DefaultServiceClient.java [hunk 2]
```diff
@@ -136,15 +134,26 @@ public void connect(final URI uri) {
     Preconditions.checkNotNull(uri, "URI must be specified.");
     Preconditions.checkArgument(uri.getScheme().equals("rosrpc"), "Invalid service URI.");
     Preconditions.checkState(tcpClient == null, "Already connected.");
-    final InetSocketAddress address = new InetSocketAddress(uri.getHost(), uri.getPort());
-    handshakeLatch.reset();
-    tcpClient = tcpClientManager.connect(toString(), address);
+    synchronized (this) {
+      Preconditions.checkState(!triedConnect, "Already called connect.");
```
Comment (lhg-turingvideo) [inline rosjava/src/main/java/org/ros/internal/node/service/DefaultServiceClient.java:138]: Even if tcpClient is null, wo should not connect again?

Comment (workingenius) [reply rosjava/src/main/java/org/ros/internal/node/service/DefaultServiceClient.java:138]: 我觉得这样处理比较简单
如果一个 client 对象允许多次调用 connect，还得继续排查 tcpClientManager 会不会造成资源泄漏
而且现在一个 client 的确只会 connect 一次
是不是

File: rosjava/src/main/java/org/ros/internal/node/service/DefaultServiceClient.java [hunk 3]
```diff
@@ -136,15 +134,26 @@ public void connect(final URI uri) {
...
+        if (!handshakeLatch.await(1, TimeUnit.SECONDS)) {
+          throw new RosRuntimeException(handshakeLatch.getErrorMessage());
+        }
+      } catch (final InterruptedException e) {
+        throw new RosRuntimeException("Handshake timed out.");
       }
-    } catch (final InterruptedException e) {
-      throw new RosRuntimeException("Handshake timed out.");
+    } finally {
+      serviceClientHandshakeHandler.shutdown();
```
Comment (huge-dicarded) [inline rosjava/src/main/java/org/ros/internal/node/service/DefaultServiceClient.java:156]: Should shutdown even if no exception?

File: rosjava/src/main/java/org/ros/internal/transport/BaseClientHandshakeHandler.java [hunk 4]
```diff
@@ -97,4 +97,8 @@ public void run(ClientHandshakeListener listener) {
       }
     });
   }
+
+  public void shutdown() {
```
Comment (huge-dicarded) [inline rosjava/src/main/java/org/ros/internal/transport/BaseClientHandshakeHandler.java:101]: Who will call this method?

**Review by huge-dicarded — CHANGES_REQUESTED** [review]
Comments added

File: rosjava/src/main/java/org/ros/internal/node/service/DefaultServiceClient.java (hunk 3 above)
Comment (workingenius) [reply rosjava/src/main/java/org/ros/internal/node/service/DefaultServiceClient.java:156]: Yes. The service handshake is only for fetching service info like host ip and port.
After that, handshake handler is useless.

File: rosjava/src/main/java/org/ros/internal/transport/BaseClientHandshakeHandler.java (hunk 4 above)
Comment (workingenius) [reply rosjava/src/main/java/org/ros/internal/transport/BaseClientHandshakeHandler.java:101]: `DefaultServiceClient` line 156 is it.

**Review by lhg-turingvideo — APPROVED** [review]

**Review by huge-dicarded — APPROVED** [review]

**Pull request merged**